keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/KDEVertex.java

```diff
@@ -0,0 +1,135 @@
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class KDEVertex extends ProbabilisticDouble {
+    public double bandwidth;
+    private DoubleTensor samples;
+
+    public KDEVertex(DoubleTensor samples, double bandwidth){
+        this.samples = samples;
+        this.bandwidth = bandwidth;
+    }
+
+    public KDEVertex(DoubleTensor samples) {
+        this.samples = samples;
+        this.bandwidth = scottsBandwith();
+    }
+
+    public KDEVertex(List<Double> samples) {
+        this(DoubleTensor.create(samples.stream()
+            .mapToDouble(Double::doubleValue)
+            .toArray()));
+    }
+
+    public KDEVertex(List<Double> samples, double bandwidth) {
+        this(DoubleTensor.create(samples.stream()
+            .mapToDouble(Double::doubleValue)
+            .toArray()), bandwidth);
+    }
+
+    public DoubleTensor pdf(DoubleTensor x) {
+        List<Double> xAsList = x.asFlatList();
+        double[] pdfs = new double[xAsList.size()];
+        for (int i = 0; i < xAsList.size(); i++){
+            DoubleTensor diff = DoubleTensor.create(xAsList.get(i), new int[]{1}).minus(samples);
+            DoubleTensor gaussian = gaussianKernel(diff.divInPlace(bandwidth)).sum(0, 1);
+            pdfs[i] = gaussian.divInPlace(DoubleTensor.create(samples.getLength() * bandwidth, gaussian.getShape())).scalar();
+        }
+        return DoubleTensor.create(pdfs);
+    }
+
+    @Override
+    public double logPdf(DoubleTensor x) {
+        return pdf(x).log().sum();
+    }
+
+    @Override
+    public Map<Long, DoubleTensor> dLogPdf(DoubleTensor value) {
+        Map<Long, DoubleTensor> partialDerivates = new HashMap<>();
+        
+        List<Double> valueAsList = value.asFlatList();
+        double[] dlnPdfs = new double[valueAsList.size()];
+        for (int i = 0; i < valueAsList.size(); i++){
+            dlnPdfs[i] = dlnPdf(DoubleTensor.create(valueAsList.get(i), new int[]{1})).dLogPdx.scalar();
+        }
+        partialDerivates.put(getId(), DoubleTensor.create(dlnPdfs, new int[]{dlnPdfs.length}));
+        return partialDerivates;
+    }
+
+    private DoubleTensor dPdx(DoubleTensor x) {
+        DoubleTensor diff = x.minus(samples).divInPlace(bandwidth);
+        return gaussianKernel(diff).times(diff).unaryMinusInPlace().sum(0, 1) .divInPlace(bandwidth).divInPlace(bandwidth * samples.getLength());
+    }
+
+    public DiffLogP dlnPdf(DoubleTensor value) {
+        return new DiffLogP(dPdx(value).divInPlace(pdf(value)));
+    }
+
+    private DoubleTensor gaussianKernel(DoubleTensor x) {
+        DoubleTensor exponent = x.pow(2.).times(-1. / 2.);
+        DoubleTensor power = DoubleTensor.create(Math.E, exponent.getShape()).powInPlace(exponent);
+        power.timesInPlace(1. / Math.sqrt(2. * Math.PI));
+        return power;
+    }
+
+    private double getMean() {
+        return samples.sum() / samples.getLength();
```

Comment:
Do we need to catch divide-by-zero error?

Reply:
also in `getVariance`

Reply:
good point, I'll instead throw an exception if an empty list of samples is provided to the constructor